### ISSUE TITLE
Add free attestations to fork choice

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#406]
## Description

Presently, only attestations that are included in blocks are included in fork choice.

They were not implemented in the fork-choice refactor (#398), instead we deferred them for later implementation.

## Steps to resolve

Attestations provided to the [`process_attestation`](https://github.com/sigp/lighthouse/blob/8afe8b3569085d05eba78a0cb11399485320405b/beacon_node/beacon_chain/src/beacon_chain.rs#L469-L502) function should be validated and provided to the [`ForkChoice`](https://github.com/sigp/lighthouse/blob/8afe8b3569085d05eba78a0cb11399485320405b/beacon_node/beacon_chain/src/fork_choice.rs) struct.

## Notes

- Only valid, signed attestations should be added to the fork choice.
- It does not suffice to only add attestations to the fork choice if they are accepted into the operations pool. The op-pool is only interested in attestations that can be included in future blocks, whilst fork choice is potentially interested in any attestation.
- Attestations should not be added to fork choice if they target unknown/unprocessed blocks.
- It may be optimal to check if the attestation is a latest message for any validator prior to verifying the signature. If the attestation is not a latest message, we don't need it and verifying the signature would be a waste of time.
- A naive implementation would read the `BeaconState` from disk for each attestation. A better implementation would attempt to use the present, canonical state prior to loading from disk. I imagine the majority of attestations would be current and could be validated against the present state, it would be good to optimize for this scenario.

**Comment by g-r-a-n-t** [comment]
Hey @paulhauner, is it cool if I pick this one up?

**Comment by paulhauner** [comment]
Please do! I think this will be a good one for you.

We're in the process of updating to the "frozen" spec (v0.8.0) but I don't think that'll affect you. In the worst case I imagine there will be a small merge conflict at some point -- nothing to worry about.

If you need anything, @michaelsproul and I are happy to help!

**Comment by vs77bb** [comment]
@g-r-a-n-t Funded this via our the EF budget and reserved for you. Please 'Start Work' whenever you can 🙂 good luck!

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 12 months from now.
Please review their action plans below:


**1) [g-r-a-n-t](https://gitcoin.co/profile/g-r-a-n-t) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/sigp/lighthouse/406/3241?mutate_worker_action=approve&worker=g-r-a-n-t) | [reject worker](https://gitcoin.co/issue/sigp/lighthouse/406/3241?mutate_worker_action=reject&worker=g-r-a-n-t))_.**

Complete what's been discussed on the ticket and WIP PR.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/sigp/lighthouse/406/3241).

**Comment by gitcoinbot** [comment]
@g-r-a-n-t Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=1) | [3 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=3) | [5 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=5) | [10 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=10) | [100 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=100)

**Comment by g-r-a-n-t** [comment]
@gitcoinbot wrapping this up

**Comment by gitcoinbot** [comment]
@g-r-a-n-t Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=1) | [3 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=3) | [5 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=5) | [10 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=10) | [100 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=100)

**Comment by g-r-a-n-t** [comment]
@gitcoinbot Waiting on a review.

**Comment by gitcoinbot** [comment]
@g-r-a-n-t Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=1) | [3 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=3) | [5 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=5) | [10 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=10) | [100 days](https://gitcoin.co/issue/sigp/lighthouse/406/3241?snooze=100)

**Comment by g-r-a-n-t** [comment]
@gitcoinbot The commits are making their way in.

**Comment by paulhauner** [comment]
@gitcoinbot please feel free to payout this issue :)

Thanks @g-r-a-n-t and @gitcoinbot!

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 550.0 DAI (550.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@g-r-a-n-t](https://gitcoin.co/profile/g-r-a-n-t)



@vs77bb please take a look at the submitted work:
* [PR](https://github.com/sigp/lighthouse/pull/437) by @g-r-a-n-t

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/sigp/lighthouse/406/3241)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/sigp/lighthouse/406/3241).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $143,994.49 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
